# Problem 3.4

This problem explores the "Gambler's Fallacy" --- the idea that after a string of losses, you are "due" for a win.

We will represent __heads__ by $+1$ and __tails__ by $-1$.  The variable $X$ is the sum of ten such flips of a coin: the number of heads minus the number of tails.

In [2]:
# Standard setup
import numpy as np
import matplotlib.pyplot as plt

# Get a random number generator.
from numpy.random import default_rng
rng = default_rng()
rand = rng.random

Verify that the following code generates a sequence of 10 flips, as described in the problem.  Briefly explain how it works.

-----

In [3]:
flips = 2*(rand(10) < 0.5) - 1 
X = flips.sum()

print("Coin flips: ", flips)
print("Sum: X=", X)

print(flips[:5])
print(sum(flips[:5]))

if sum(flips[:5]) == 5:
      print("The sum of the first 5 elements was 5!")
else: print("The sum of the first 5 elements was NOT 5!")

Coin flips:  [-1 -1 -1  1 -1 -1 -1  1 -1  1]
Sum: X= -4
[-1 -1 -1  1 -1]
-3
The sum of the first 5 elements was NOT 5!


The first line creates a boolean array of 10 randomly generated numbers that fit a 'True/False' statement if < 0.5 (or 0 = False and 1 = True) and multiplies that value it by 2, then is subtracts 1 from the resulting product to give us an array of consisting of 1 for True and -1 for False. Then the second line of code sums together all of the values in the matrix labeled 'flips' and assigns this sum to the variable 'X'. The third and fourth lines of code prints the first two lines of code with short descriptions. The fifth and sixth line of code prints the first 5 flips of a coin and the sum of these 5 flips, respectively. Finally, the last if/else statement generates a final print line that will print 'The sum of the first 5 elements was 5!' if the sum of the first 5 flips really does sum to 5, or else it prints 'The sum of the first 5 elements was NOT 5!'

------

## (a)

Fill in the necessary Python code below.

In [29]:
# Define the problem.
Ntrials = 200
Nflips = 10 
Nstart = 5

# Create an array to store the data we want.
data = np.zeros(Ntrials)
count = 0

# Generate Ntrials sequences of Nflips coin flips, and sum.
for n in range(Ntrials):
    flips = 2*(rand(Nflips) < 0.5) - 1 #rand(Nflips)
    X = flips.sum()
    
    # Replace "if True:" with a conditional statement that will do the following.
    #    IF the first Nstart flips are heads, store X in "data" and increment "count".
    #    OTHERWISE, do nothing and continue to the next trial.
    if sum(flips[:5]) == 5:
        data[count] = X
        count += 1 #means the same thing as "count = count + 1" the += is the count times assignment

# Now use the data set to answer the question.
if count == 0:
    print("No trials started with {} heads.".format(Nstart))
else:
    Num = data[:count].sum()
    avgX = Num / count
    print ("There were {} trials that started with {} heads.".format(count,Nstart))
    print("The average of X for these trials was <X> = {:.4f}.".format(avgX))

There were 12 trials that started with 5 heads.
The average of X for these trials was <X> = 4.6667.


Run your code several times.  What patterns do you observe?

-----

When there are 200 trials of 10 flips, I have noticed there is a pattern of very few of these trials starting off with the first 5 flips ending up landing on Heads. Out of these very few trials, typically ranging from 3 to 12 trials starting with 5 Heads out of the total 200 trials, the average sum (which we assume wants to be as close to 0 as possible) of the 10 flips in each trial ranges between 3 and 6. 

-----

## (b)

Repeat for `Ntrials = 2000` and `Ntrials = 8000`.

> Does the answer seem to be converging to zero, as predicted in the quoted text above? Whatever your answer, give some explanation for why the answer you found should have been expected.

-----

In [59]:
# Define the problem.
Ntrials = 8000
Nflips = 10 
Nstart = 5

# Create an array to store the data we want.
data = np.zeros(Ntrials)
count = 0

# Generate Ntrials sequences of Nflips coin flips, and sum.
for n in range(Ntrials):
    flips = 2*(rand(Nflips) < 0.5) - 1 #rand(Nflips)
    X = flips.sum()
    
    # Replace "if True:" with a conditional statement that will do the following.
    #    IF the first Nstart flips are heads, store X in "data" and increment "count".
    #    OTHERWISE, do nothing and continue to the next trial.
    if sum(flips[:5]) == 5:
        data[count] = X
        count += 1 #means the same thing as "count = count + 1" the += is the count times assignment

# Now use the data set to answer the question.
if count == 0:
    print("No trials started with {} heads.".format(Nstart))
else:
    Num = data[:count].sum()
    avgX = Num / count
    print ("There were {} trials that started with {} heads.".format(count,Nstart))
    print("The average of X for these trials was <X> = {:.4f}.".format(avgX))

There were 231 trials that started with 5 heads.
The average of X for these trials was <X> = 5.0043.


The average sum of these trials of 10 coin flips that start with the first 5 flips landing on heads actually seems to converge to 5 rather than 0, like we origanlly expected. This makes more sense because each coin flip is independantly probable to be 50% chance Heads and 50% chance Tails, not dependant upon the fact that the first five flips were all Heads so the second five flips would have to be tails. Therefore, we can't assume that the sum of the second set of 5 flips is = -5 but rather closer to 0 because each flip is independantly 50/50 chance heads/tails. So it would make sense that with an initial sum of 5, the variance from there out is 50/50 and closer to an average sum of X = 0 for the second set of 5 flips, and a total average sum of X = 5 for the total 10 flips.

-----

## (c)

Explore regression to the mean, as described in the problem text.

In [73]:
# Practice example of how to change the number of flips per trial
Ntrials = 200
Nvalues = [10, 100, 500, 2500, 50000] 
Nflips = 10
Nstart = 5

for n in Nvalues:
    # Create an array to store the data we want.
    data = np.zeros(Ntrials)
    count = 0
    print('n= ',n)
    Nflips = n

    # Generate Ntrials sequences of Nflips coin flips, and sum.
    for n in range(Ntrials):
        flips = 2*(rand(Nflips) < 0.5) - 1 
        X = flips.sum()
    
        # Replace "if True:" with a conditional statement that will do the following.
        #    IF the first Nstart flips are heads, store X in "data" and increment "count".
        #    OTHERWISE, do nothing and continue to the next trial.
        if sum(flips[:5]) == 5:
            data[count] = X
            count += 1 #means the same thing as "count = count + 1" the += is the count times assignment

    # Now use the data set to answer the question.
    if count == 0:
        print("No trials started with {} heads.".format(Nstart))
    else:
        avgX = data[:count].sum() / count
        print ("There were {} trials that started with {} heads.".format(count,Nstart))
        print("The average of X for these trials was <X>= {:.4f}.".format(avgX))


n=  10
There were 5 trials that started with 5 heads.
The average of X for these trials was <X>= 4.4000.
n=  100
There were 4 trials that started with 5 heads.
The average of X for these trials was <X>= -3.0000.
n=  500
There were 1 trials that started with 5 heads.
The average of X for these trials was <X>= -16.0000.
n=  2500
There were 8 trials that started with 5 heads.
The average of X for these trials was <X>= -13.2500.
n=  50000
There were 3 trials that started with 5 heads.
The average of X for these trials was <X>= 68.6667.


Comment on your observations.

-----

I noticed that as the number of times you flip the coin in each trial increases, the regression towards the mean also increases and has a very large variance. Therefore a good pattern to take away from this is that the more flips you consider in each trial causes a less accurate representation of the probability of the first 5 flips being heads, because the variance in the mean is increasingly larger the more flips considered in each trial; however, the increasing number of trials of 10 coint flips taken into consideration will only give a more accurate representation of what is actually the probability of the first five flips in a trial being heads.

-----

## (d)

Evaluate the ___spread___ of $X$ for large numbers of trials, as described in the text.

In [89]:
X = rand(1000)
print("X= {:.4f} +/- {:4f}".format(X.mean(), X.std()))

#or
np.std(X)

X= 0.4981 +/- 0.287130


0.28712959256420156

Comment on your observations.

-----

The spread of X for large numbers of trials is actually very small. The spread of this trial of 1000 random number has a standard deviation or a spread that typically converges to 0.29

_____

## (e)

Discuss which proposition better describes your findings.

> If the first N flips are heads, the next series flips will be more than half tails, in order to pull X closer to zero, because X ‘wants’ to be zero on average.

... or ...

> The effect of unusual past behavior doesn’t disappear; it just gets diluted as time goes on.

-----

The effects of unusual past behavior doesn't disappear; it just get diluted as time goes on is a better representation of my findings because in the first part of the problem, when an abnormal event happens such as the first five flips of a 10 flip trial of a coin toss land on heads, the average sum of the total 10 flips in the trial converged to 5, meaning the effects of the abnormal event stayed around rather than pulled back to equilibrium. However, as time goes on, or as more and more trials are taken, the effects of the first five flips being heads gets diluted due to the normal events of the rest of the coin flip averages converging to zero.